### ISSUE TITLE
Heap use after free memory 

### ISSUE THREAD

**Issue by altavist** [Cylix/cpp_redis#74]
Hello together,

first of all thanks for this well written C++ redis library code. 

I integrated your library into my own private repository. After playing with it, I detect some issue in case of the standard disconnection handling. 

The class _redis_connection_ contains the function _disconnect_ with a "wait_for_removal" flag which is predefined as false. 

`  void disconnect(bool wait_for_removal = false);
`

The class _redis_client_ has got a deconstructor which is using this function with the "wait_for_removal" flag predefined as true.

`redis_client::~redis_client(void) 
{
   m_client.disconnect(true);  /* parameter = true     */
  __CPP_REDIS_LOG(debug, "cpp_redis::redis_client destroyed");
}
`

The class redis_client also uses the mentioned function above in such a way that this flag is false:

`void
redis_client::disconnect(void) {
  __CPP_REDIS_LOG(debug, "cpp_redis::redis_client attempts to disconnect");
  m_client.disconnect(); /* no parameter = false      */
__CPP_REDIS_LOG(info, "cpp_redis::redis_client disconnected");
}`

So when I´m explicitly using the disconnect-function from a redis_client object, the wait_for_removal flag is always false. Because of this, I get a "heap use after free" memory issue while I was connecting / disconnecting repeatedly. In the issue case, the disconnection has been closed without waiting for removal. While this happened, an async_read gets executed and the error occured. 
 
Check my address sanitizer logfile for further information. 

Is this expected behavior from your side? Am I using the disconnection function wrong or did you forget setting the flag predefined as true?

Thanks in advance,

Altavist
 
[asan.txt](https://github.com/Cylix/cpp_redis/files/970517/asan.txt)

**Comment by Cylix** [comment]
Hi,

Indeed, wait_for_removal will hang until the underlying socket has been removed from the IO_service (event loop) and that all currently executing callbacks are done.

At first, I decided to let `disconnect` use the default `false` behavior for performance issues (mostly if someone want to disconnect asynchronously and do something at the same time), but I guess it causes some issues when `disconnect` is one of the last instruction of the program.

I'll update that tonight by providing an optional parameter to `disconnect` in the same way as there is an optional `wait_for_removal` in the `redis_connection`.
I think it is going to be false by default to avoid breaking change, but this way you will be able to override the default setting.

Thanks for opening this issue btw :)

Best!
